### PR TITLE
Add nftables resources

### PR DESCRIPTION
**Pull request by jjacque**
## Description
[nftables](https://netfilter.org/projects/nftables/) is a netfilter project that aims to replace the existing {ip,ip6,arp,eb}tables framework.
This PR add the `nftables` resource that allows you to check:
- chain attributes (`type`, `hook`, `prio`, `policy`)
- rules existence in a given chain
- named set attributes (`type`, `flags`, `size`)
- element existence in a given named set

Usage of `nftables` instead of `iptables` in the kitchen runs is enabled using a node attributes `['osprepare']['nftables']`, `false` by default. Tests are activated by the presence of the `NFTABLES` env variable.
When `nftables` recipes/tests are enable the `iptables` one is disabled.

I also forced `chef-client` version to `17` in kitchen.dokken as the [docker](https://supermarket.chef.io/cookbooks/docker) cookbook needs chef-client < 18.

This PR was inspired by https://github.com/inspec/inspec/pull/4236.

## Related Issue

https://github.com/inspec/inspec/issues/4034

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Review by Vasu1105 — APPROVED** [review]

**Comment by Vasu1105** [comment]
This LGTM. Thanks @jjacque.

**Comment by jjacque** [comment]
Hi @Vasu1105 , i just added a commit to remove the `-nn` nft opt during chain rules retrieval as based on my testing on Ubuntu 22.04 the behaviour changed.
Thanks for your review.

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    describe nftables(family:'name', table:'name', chain: 'name') do
+      its('PROPERTY') { should eq 'value' }
+      it { should have_rule('RULE') }
+    end
+    o
+    describe nftables(family:'name', table:'name', set: 'name') do
+      its('PROPERTY') { should eq 'value' }
+      it { should have_element('ELEMENT') }
+    end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
```ruby
describe nftables(family:'name', table:'name', chain: 'name') do
  its('PROPERTY') { should eq 'value' }
  it { should have_rule('RULE') }
end

describe nftables(family:'name', table:'name', set: 'name') do
  its('PROPERTY') { should eq 'value' }
  it { should have_element('ELEMENT') }
end
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+## Availability
+
+### Installation
+
+{{% inspec/inspec_installation %}}
+
+### Version
+
+This resource first became available in v4.6.9 of InSpec.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]: Version number?

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      its('hook') { should eq 'input' }
+      its('prio') { should eq (-150) } # maangle
+      its('policy') { should eq 'accept' }
+    end
+
+### Test if there is a rule allowing Postgres (5432/TCP) traffic
+
+    describe nftables(family: 'inet', table: 'my_filter', chain: 'my_input') do
+      it { should have_rule('tcp dport 5432 comment "postgres" accept') }
+    end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
```ruby
describe nftables(family: 'inet', table: 'TABLE_NAME', chain: 'CHAIN_NAME') do
  it { should have_rule('tcp dport 5432 comment "postgres" accept') }
end
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+Note that the rule specification must exactly match what's in the output of `nftables -nn list chain inet my_filter my_input`, which will depend on how you've built your rules.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
```suggestion
Note that the rule specification must exactly match what's in the output of `nftables -nn list chain inet TABLE_NAME CHAIN_NAME`, which will depend on how you've built your rules.
```

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+### Test if there is an element `1.1.1.1` in the OPEN_PORTS named set
+
+    describe nftables(family: 'inet', table: 'my_filter', set: 'OPOEN_PORTS') do
+      it { should have_element('1.1.1.1') }
+    end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
### Test if there is an element `1.1.1.1` in the OPEN_PORTS named set

```ruby
describe nftables(family: 'inet', table: 'TABLE_NAME', set: 'OPEN_PORTS') do
  it { should have_element('1.1.1.1') }
end
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+## Matchers
+
+For a full list of available matchers, please visit our [matchers page](/inspec/matchers/).
+
+### have_rule
+
+The `have_rule` matcher tests the named rule against the information in the `nftables` ruleset:
+
+    it { should have_rule('RULE') }
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
```ruby
it { should have_rule('RULE') }
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    end
+
+### Test if the INPUT chain from the mangle table is in ACCEPT mode
+
+    describe nftables(family: 'inet', table: 'mangle', chain: 'INPUT') do
+      its('type') { should eq 'filter' }
+      its('hook') { should eq 'input' }
+      its('prio') { should eq (-150) } # maangle
+      its('policy') { should eq 'accept' }
+    end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
### Test if the `CHAIN_NAME` chain from the `TABLE_NAME` table has the default `accept` policy

```ruby
describe nftables(family: 'inet', table: 'TABLE_NAME', chain: 'CHAIN_NAME') do
  its('type') { should eq 'filter' }
  its('hook') { should eq 'input' }
  its('prio') { should eq (-150) }
  its('policy') { should eq 'accept' }
end
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+The `have_rule` matcher tests the named rule against the information in the `nftables` ruleset:
+
+    it { should have_rule('RULE') }
+
+### have_element
+
+The `have_element` matcher tests the named set against the information in the `nftables` ruleset:
+
+    it { should have_element('SET_ELEMENT') }
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
```ruby
it { should have_element('SET_ELEMENT') }
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+## Examples
+
+The following examples show how to use this Chef InSpec audit resource.
+
+### Test if the my_input chain is in default ACCEPT mode
+
+    describe nftables(family: 'inet', table: 'my_filter', chain: 'my_input') do
+      its('policy') { should eq 'accept' }
+    end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
````suggestion
### Test if the `CHAIN_NAME` chain has the default `accept` policy

```ruby
describe nftables(family: 'inet', table: 'TABLE_NAME', chain: 'CHAIN_NAME') do
  its('policy') { should eq 'accept' }
end
```
````

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+where
+
+- `nftables()` has to specify `family` and `table`. It also has to specify one of `chain` or `set` (exclusively).
+- `family:'name'` is the name of the `family` the table belongs to, one of `ip`, `ip6`, `inet`, `arp`, `bridge`, `netdev`.
+- `table:'name'` is the packet matching table against which the test is run.
+- `chain: 'name'` is the name of a user-defined chain.
+- `set: 'name'` is the name of a user-defined named set.
+- `have_rule('RULE')` tests that the chain has a given rule in the nftables ruleset. This must match the entire line taken from `nftables -nn list chain FAMILY TABLE CHAIN`.
+- `have_element('ELEMENT')` tests that element is a member of the nftables named set.
+
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md:51]:
```suggestion

See the [NFT man page](https://www.netfilter.org/projects/nftables/manpage.html) and [nftables wiki](https://wiki.nftables.org/wiki-nftables/index.php/Main_Page) for more information about nftables.

```

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+- `hook`
+- `prio`
+- `policy`
+- `type`
+
+### Set Properties
+
+- `flags`
+- `size`
+- `type`
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
```suggestion
`flags`
: The set flags. Possible values: `constant`, `dynamic`, `interval`, and `timeout`.
`size`
: The maximum number of elements in the set.
`type`
: The data type of set elements. Possible values: `ipv4_addr`, `ipv6_addr`, `ether_addr`, `inet_proto`, `inet_service`, and `mark`. 
```

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+- `have_element('ELEMENT')` tests that element is a member of the nftables named set.
+
+## Properties
+
+### Chain Properties
+
+- `hook`
+- `prio`
+- `policy`
+- `type`
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
```suggestion
`hook`
: The hook type. Possible values: `ingress`, `prerouting`, `forward`, `input`, `output`, `postrouting`, and `egress`.
`prio`
: The numerical chain priority.
`policy`
: The policy type. Possible values: `accept`, `drop`.
`type`
: The chain type.  Possible values: `filter`, `nat`, and `route`.
```

File: docs-chef-io/content/inspec/resources/nftables.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+platform = "linux"
+
+[menu]
+  [menu.inspec]
+    title = "nftables"
+    identifier = "inspec/resources/os/nftables.md nftables resource"
+    parent = "inspec/resources/os"
++++
+
+Use the `nftables` Chef InSpec audit resource to test rules and sets that are defined using `nftables`, which maintains tables of IP packet filtering rules. There may be more than one table. Each table contains one (or more) chains. A chain is a list of rules that match packets. When the rule matches, the rule defines what target to assign to the packet.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/nftables.md]:
```suggestion
Use the `nftables` Chef InSpec audit resource to test rules and sets that are defined using `nftables`, which maintains tables of IP packet filtering rules. There may be more than one table. Each table contains one (or more) chains. A chain is a list of rules that match packets. When a rule matches a packet, the rule defines what target to assign to the packet.
```

**Review by IanMadd — COMMENTED** [review]
A bunch of docs feedback/suggestions/edits.

**Comment by jjacque** [comment]
Hi @IanMadd , thanks for your comments, just fixed it.

**Review by IanMadd — APPROVED** [review]

**Pull request merged**